### PR TITLE
Display signup/login form

### PR DESCRIPTION
**Pull request by davidboll**
#### Description
Changes proposed in this pull request:

* adds signup and login function 

#### What I have learned working on this feature: 
How to implement devise 

### Screenshots:
screen shot 2018-02-27 at 4 31 09 pm
screen shot 2018-02-27 at 4 31 14 pm

### CODE REVIEW

File: spec/models/user_spec.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+  pending "add some examples to (or delete) #{__FILE__}"
```
Comment (faraznaeem) [inline spec/models/user_spec.rb]: Remove if not in use

Comment (tochman) [reply spec/models/user_spec.rb]: Or add tests. I don’t understand why there are no unit tests for User class

File: spec/user_spec.rb [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+RSpec.describe User do
+
+  describe "Factory" do
+    it "should have a valid factory" do
+      expect(FactoryBot.create(:user)).to be_valid
+    end
+  end
+
+end
```
Comment (tochman) [inline spec/user_spec.rb]: This spec should be in the models folder

**Review by tochman — CHANGES_REQUESTED** [review]
Why did you generate all devise views?

File: app/helpers/welcome_helper.rb [hunk 3]
```diff
@@ -0,0 +1,2 @@
+module WelcomeHelper
```
Comment (diraulo) [inline app/helpers/welcome_helper.rb]: You don't need this helper file at this stage and chance are you'll likely not add any helper here so remove this file

File: app/views/devise/mailer/email_changed.html.erb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+<p>Hello <%= @email %>!</p>
```
Comment (diraulo) [inline app/views/devise/mailer/email_changed.html.erb]: Unless you're overriding the default behavior of Devise you do not need to generate this view. So delete this file

File: app/views/devise/mailer/password_change.html.erb [hunk 5]
```diff
@@ -0,0 +1,3 @@
+<p>Hello <%= @resource.email %>!</p>
```
Comment (diraulo) [inline app/views/devise/mailer/password_change.html.erb]: Same here you don't need this

File: app/views/devise/mailer/reset_password_instructions.html.erb [hunk 6]
```diff
@@ -0,0 +1,8 @@
+<p>Hello <%= @resource.email %>!</p>
```
Comment (diraulo) [inline app/views/devise/mailer/reset_password_instructions.html.erb]: Delete this..

File: app/views/devise/mailer/unlock_instructions.html.erb [hunk 7]
```diff
@@ -0,0 +1,7 @@
+<p>Hello <%= @resource.email %>!</p>
```
Comment (diraulo) [inline app/views/devise/mailer/unlock_instructions.html.erb]: Delete this

File: app/views/devise/passwords/edit.html.erb [hunk 8]
```diff
@@ -0,0 +1,25 @@
+<h2>Change your password</h2>
```
Comment (diraulo) [inline app/views/devise/passwords/edit.html.erb]: Delete this

File: app/views/devise/passwords/new.html.erb [hunk 9]
```diff
@@ -0,0 +1,16 @@
+<h2>Forgot your password?</h2>
```
Comment (diraulo) [inline app/views/devise/passwords/new.html.erb]: Delete this

File: app/views/devise/registrations/edit.html.erb [hunk 10]
```diff
@@ -0,0 +1,43 @@
+<h2>Edit <%= resource_name.to_s.humanize %></h2>
```
Comment (diraulo) [inline app/views/devise/registrations/edit.html.erb]: Delete this one too

File: app/views/devise/registrations/new.html.erb [hunk 11]
```diff
@@ -0,0 +1,29 @@
+<h2>Sign up</h2>
```
Comment (diraulo) [inline app/views/devise/registrations/new.html.erb]: Delete this one too

File: app/views/devise/sessions/new.html.erb [hunk 12]
```diff
@@ -0,0 +1,26 @@
+<h2>Log in</h2>
```
Comment (diraulo) [inline app/views/devise/sessions/new.html.erb]: Delete this one too

File: app/views/devise/shared/_links.html.erb [hunk 13]
```diff
@@ -0,0 +1,25 @@
+<%- if controller_name != 'sessions' %>
```
Comment (diraulo) [inline app/views/devise/shared/_links.html.erb]: Delete this one too

File: app/views/devise/unlocks/new.html.erb [hunk 14]
```diff
@@ -0,0 +1,16 @@
+<h2>Resend unlock instructions</h2>
```
Comment (diraulo) [inline app/views/devise/unlocks/new.html.erb]: Delete this one too

File: app/views/layouts/application.html.erb [hunk 15]
```diff
@@ -4,11 +4,21 @@
...
-    <%= yield %>
+    <%= render 'layouts/nav' %>
+
+    <% flash.each do |key, value| %>
+      <div class="text-center">
+        <%= value %>
+      </div>
+    <% end %>
+
+  <%= yield %>
```
Comment (diraulo) [inline app/views/layouts/application.html.erb]: Fix indentation of this line

File: config/environments/development.rb [hunk 16]
```diff
@@ -1,54 +1,23 @@
 Rails.application.configure do
-  # Settings specified here will take precedence over those in config/application.rb.
```
Comment (diraulo) [inline config/environments/development.rb:2]: Making changes like this just add too much clutter in the diff of the PR and makes it hard to review actual changes that are relevant to the PR itself. I get that you may have deleted these comments in response to a previous comment about having too many changes in the PR, but you have to understand that these comments were not counted as part of the changes of this PR but now they are because you removed them. 

Will let this slide for now but in future only modify files that directly have an impact on your feature

File: config/environments/production.rb [hunk 17]
```diff
@@ -1,91 +1,22 @@
 Rails.application.configure do
-  # Settings specified here will take precedence over those in config/application.rb.
```
Comment (diraulo) [inline config/environments/production.rb:2]: Same this for this one

File: config/environments/test.rb [hunk 18]
```diff
@@ -1,42 +1,15 @@
 Rails.application.configure do
-  # Settings specified here will take precedence over those in config/application.rb.
```
Comment (diraulo) [inline config/environments/test.rb:2]: and this one

File: config/initializers/application_controller_renderer.rb [hunk 19]
```diff
@@ -1,8 +0,0 @@
-# Be sure to restart your server when you modify this file.
```
Comment (diraulo) [inline config/initializers/application_controller_renderer.rb]: Please revert the changes on this file

File: config/initializers/assets.rb [hunk 20]
```diff
@@ -1,14 +1,2 @@
-# Be sure to restart your server when you modify this file.
```
Comment (diraulo) [inline config/initializers/assets.rb]: Revert changes on this file too

File: config/initializers/backtrace_silencers.rb [hunk 21]
```diff
@@ -1,7 +0,0 @@
-# Be sure to restart your server when you modify this file.
```
Comment (diraulo) [inline config/initializers/backtrace_silencers.rb]: Revert changes on this file

File: config/initializers/cookies_serializer.rb [hunk 22]
```diff
@@ -1,5 +1 @@
-# Be sure to restart your server when you modify this file.
```
Comment (diraulo) [inline config/initializers/cookies_serializer.rb]: Revert changes here

File: config/initializers/filter_parameter_logging.rb [hunk 23]
```diff
@@ -1,4 +1 @@
-# Be sure to restart your server when you modify this file.
```
Comment (diraulo) [inline config/initializers/filter_parameter_logging.rb]: Revert the changes on this file too

File: config/initializers/inflections.rb [hunk 24]
```diff
@@ -1,16 +0,0 @@
-# Be sure to restart your server when you modify this file.
```
Comment (diraulo) [inline config/initializers/inflections.rb]: revert changes on this file as well

File: config/initializers/mime_types.rb [hunk 25]
```diff
@@ -1,4 +0,0 @@
-# Be sure to restart your server when you modify this file.
```
Comment (diraulo) [inline config/initializers/mime_types.rb]: and this one

File: config/initializers/wrap_parameters.rb [hunk 26]
```diff
@@ -1,14 +1,3 @@
-# Be sure to restart your server when you modify this file.
```
Comment (diraulo) [inline config/initializers/wrap_parameters.rb]: and this one as well

File: config/locales/en.yml [hunk 27]
```diff
@@ -1,33 +1,2 @@
-# Files in the config/locales directory are used for internationalization
```
Comment (diraulo) [inline config/locales/en.yml]: Why are we modifying this file? 🤔

File: db/schema.rb [hunk 28]
```diff
@@ -1,18 +1,22 @@
-# This file is auto-generated from the current state of the database. Instead
```
Comment (diraulo) [inline db/schema.rb:1]: Like the comment says this is *auto-generated* so there's no point in removing these comments since they'll likely be back later on. You should never *manually* edit the schema file of your rails application. Running the migrations will take care of that

File: db/seeds.rb [hunk 29]
```diff
@@ -1,7 +0,0 @@
-# This file should contain all the record creation needed to seed the database with its default values.
```
Comment (diraulo) [inline db/seeds.rb]: Again, why was this file modified?

File: spec/rails_helper.rb [hunk 30]
```diff
@@ -13,4 +14,10 @@
 
   config.infer_spec_type_from_file_location!
   config.filter_rails_from_backtrace!
+  Shoulda::Matchers.configure do |config|
```
Comment (diraulo) [inline spec/rails_helper.rb]: This setup block for Shoulda matcher should be set outside of the `RSpec.configure` block

**Review by diraulo — DISMISSED** [review]
You do not need to generate devise views in your application unless you want to make changes to the way they look or add new fields to the various forms. In your case you're not doing any of that so please remove all devise generated views unless you changed a field on a form

All the changes you made (removing comments) in the `initializers` files have to be reverted. 

See the rest of the comments in respective files

File: app/views/layouts/application.html.erb [hunk 31]
```diff
@@ -4,11 +4,21 @@
     <title>Evensteven</title>
     <%= csrf_meta_tags %>
 
-    <%= stylesheet_link_tag    'application', media: 'all', 'data-turbolinks-track': 'reload' %>
+    <%= stylesheet_link_tag 'application', media: 'all', 'data-turbolinks-track': 'reload' %>
+    <%= stylesheet_link_tag '//maxcdn.bootstrapcdn.com/font-awesome/4.3.0/css/font-awesome.min.css' %>
```
Comment (tochman) [inline app/views/layouts/application.html.erb]: If you use font awesome, please add it using the gem. Serving assets through CDN might cause unforeseen behavior at times and is considered bad practice.

File: app/views/layouts/_nav.html.erb [hunk 32]
```diff
@@ -0,0 +1,11 @@
...
+ <ul class=“nav navbar-nav pull-right”>
+   <% if user_signed_in? %>
+     <li><%= link_to "Hello, #{current_user.email}", "#" %></li>
+     <li><%= link_to 'Logout', destroy_user_session_path, method: :delete %></li>
+   <% else %>
+     <li class=“active”> <%= link_to 'Login', new_user_session_path %> </li>
+     <li> <%= link_to 'Sign up', new_user_registration_path %> </li>
+   <% end %>
+ </ul>
+</div>
```
Comment (tochman) [inline app/views/layouts/_nav.html.erb]: I see you are using Bootstrap classes here, but I can’t see that you are actually adding Bootstrap as a dependency. Why?

**Review by tochman — DISMISSED** [review]

File: config/environments/development.rb (hunk 16 above)
Comment (davidboll) [reply config/environments/development.rb:2]: Roger that :)

File: db/seeds.rb (hunk 29 above)
Comment (davidboll) [reply db/seeds.rb]: On it 👍

File: spec/rails_helper.rb (hunk 30 above)
Comment (davidboll) [reply spec/rails_helper.rb]: Tnx!

File: app/views/layouts/application.html.erb [hunk 33]
```diff
@@ -4,11 +4,21 @@
     <title>Evensteven</title>
     <%= csrf_meta_tags %>
 
-    <%= stylesheet_link_tag    'application', media: 'all', 'data-turbolinks-track': 'reload' %>
+    <%= stylesheet_link_tag 'application', media: 'all', 'data-turbolinks-track': 'reload' %>
+    <%= stylesheet_link_tag href="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0/css/bootstrap.min.css" %>
```
Comment (tochman) [inline app/views/layouts/application.html.erb]: Again, servig assets over CDN is not a practice we encourage. Also, in case of Bootstrap, you need the js files and have jQuery installed.

File: spec/models/user_spec.rb [hunk 34]
```diff
@@ -0,0 +1,16 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+
+  describe "Factory" do
+    it "should have a valid factory" do
+      expect(FactoryBot.create(:user)).to be_valid
```
Comment (tochman) [inline spec/models/user_spec.rb]: I see that you included `FactoryBot::Syntax::Methods` in your RSpec configuration. That is really good. Noww, make sure you use those methods by shortening this line to:
``` 
expect(create(:user)).to be_valid
```

File: spec/support/factory_bot.rb [hunk 35]
```diff
@@ -0,0 +1,5 @@
+require 'factory_bot'
+
+RSpec.configure do |config|
+  config.include FactoryBot::Syntax::Methods
```
Comment (tochman) [inline spec/support/factory_bot.rb:4]: Good stuff. You can do the same for your cucumber configuration, and make use of a shorter syntax (meaning you can omit `FactoryBot` - see example above.)

File: app/views/layouts/application.html.erb (hunk 33 above)
Comment (holgertidemand) [reply app/views/layouts/application.html.erb]: I dont really understand. You guys told us to use CDN's and we are not using Js or JQuery at this point°

Comment (holgertidemand) [reply app/views/layouts/application.html.erb]: @tochman

File: spec/support/factory_bot.rb (hunk 35 above)
Comment (holgertidemand) [reply spec/support/factory_bot.rb:4]: Not sure what you mean! remove ``` require 'factory_bot``` ?

File: app/views/layouts/application.html.erb (hunk 33 above)
Comment (tochman) [reply app/views/layouts/application.html.erb]: When and who told you that? Also, I'm telling you a different thing now. Also, Bootstrap simply won't work without the js dependencies and jQuery.

Comment (holgertidemand) [reply app/views/layouts/application.html.erb]: Well we did a minihomework about why it's good and why we should use it!
But just to clarify, we should download the bootstrap lib to our project?!

File: spec/support/factory_bot.rb (hunk 35 above)
Comment (tochman) [reply spec/support/factory_bot.rb:4]: Well, now that you mention it... I would try it...
But no, I mean that in your tests, you can omit the word `FactoryBot` (as I showed in an example above) 
Out of curiosity, since you have to ask, why did you guys add this setting in the first place?

File: app/views/layouts/application.html.erb (hunk 33 above)
Comment (tochman) [reply app/views/layouts/application.html.erb]: There is probably a gem for that. Like for most other things...

**Review by tochman — DISMISSED** [review]
Still not ready to go IMO. See comments.

File: app/views/layouts/application.html.erb (hunk 33 above)
Comment (holgertidemand) [reply app/views/layouts/application.html.erb]: Thank you @tochman

**Comment by tochman** [comment]
A general comment is that a lot of the issues we've had with this PR would have been avoided if 
1. Previous PR's was in good standing
2. The devs stayed in scope (one example is adding a css framework and/or functionality for account management) 

What I see in PivotalTracker is this: 
skarmavbild 2018-02-28 kl 10 51 51
What I see in the code (even after cleaning) is way more. 
Having said that, thank you for responding to our review comments and fixing this. Code reviews are hard work. Especially during a Coding Bootcamp.

**Review by diraulo — APPROVED** [review]
There are still a couple of things I'm not entirely happy about here but I will let them slide this time around in the hope that you do a better job at this in future 🙂

**Comment by tochman** [comment]
Please do not dismiss my reviews. ;-) 
skarmavbild 2018-02-28 kl 11 56 18

**Comment by holgertidemand** [comment]
Thank you @diraulo. Can you give us a list of the things you let slide this time :)

**Comment by holgertidemand** [comment]
Sorry Thomas!
